text/0000-ember-data-identifiers.md

```diff
@@ -0,0 +1,658 @@
+#### Handling Updates to Alternative Cache Keys
+ 
+ Note that in our above example we treat `username` a stable, immutable alternative 
+  primary-key. Some APIs allow users to change the value of such "unique keys" (`email`
+  `phone` `username` being common examples).
+
+ If your application enables such behavior, and these updates are not handled by the
+  call to `updateRecordIdentifier` that occurs after a record is saved, in addition
+  to manually calling `identifierCache.updateRecordIdentifier` with the desired patch
+  you could also provide your own explicit method for doing so. An explicit method is
+  great for ensuring the correct context for the granularity of this change. The timing
+  of this update would be up to you (whether pre- or post- the mutation having been
+  persisted to the server)
+
+ Extending the example above:
+
+ ```ts
+ import { SECONDARY_IDENTIFIER_CACHE } from './initializers/configure-identifiers';
+
+ export function updateUsernameForIdentifier(
+   // application instance is the result of `getOwner`
+   // on something like the `store` or a `component`
+   owner: Owner, 
+   identifier: Identifier,
+   oldUsername: string,
+   newUsername: string
+   ) {
+    const caches = SECONDARY_IDENTIFIER_CACHE.get(applicationInstance);
+    const typeid_cache = caches.typeid;
+    const username_cache = caches.username;
+
+    if (username_cache[oldUsername] !== identifier.lid) {
+      throw new Error('invalid update');
+    }
+
+    // you might want to continue mapping both old an new username
+    //  here we decided not to.
+    delete username_cache[oldUsername];
+
+    username_cache[newUsername] = identifier.lid;
+  }
+ ```
+
+ ### Identifier Stability
+
+Identifiers handed to public APIs by `ember-data` will **always** be _referentially stable_
+ Public `ember-data` APIs that **expect** an `Identifier` will normalize the object they 
+ are given into the stable `Identifier` if it is not one already. This is done to allow
+ for serialized identifiers and identifying information from the API to more easily be worked
+ with without extra normalization effort.
+
+Specifically, this means that as regards this RFC, `identifiers` that are structurally the same
+ (meaning an object with the same `lid`) are treated as any other `identifier` regardless of
+ whether the object is an identical reference to the object generated by the store previously.
+
+Additionally, only the objects generated by the store will have additional debug information
+ attached to them, as shown in the interfaces below.
+
+```typescript
+const IS_IDENTIFIER = Symbol('is-identifier');
+
+// provided for additional debuggability
+const DEBUG_CLIENT_ORIGINATED = Symbol('record-originated-on-client');
+const DEBUG_IDENTIFIER_BUCKET = Symbol('identifier-bucket');
+
+export interface StableIdentifier {
+  lid: string;
+  [IS_IDENTIFIER]: true;
+  [DEBUG_IDENTIFIER_BUCKET]: string;
+}
+
+export interface RecordIdentifier extends StableIdentifier {
```

Comment:
Shouldn't this be `StableRecordIdentifier` ?

Reply:
Thanks for the catch! Over eager refactor 😂